### PR TITLE
Cache backend symbols in the unified backend

### PR DESCRIPTION
**Pull request by umar456**
unified

Test code:

```cpp
#include <arrayfire.h>
#include <cstdio>
#include <cstdlib>
#include <chrono>

using namespace af;

int main(int argc, char *argv[])
{
    try {
      int backends = af::getAvailableBackends();
      af::setBackend(AF_BACKEND_CUDA);
      af::info();

      printf("\nRunning %s...\n", argv[0]);

      for(int i = 500; i < 100000; i+= 500) {
        auto start = std::chrono::high_resolution_clock::now();
        for(int iter = 0; iter < 10000; iter++) {
          af::array blah = af::randu(i);
          af::array constant_val = af::constant(1, i);
          af::array output = dot(blah, constant_val);
        }
        af::sync();
        auto end = std::chrono::high_resolution_clock::now();
        af::deviceGC();
        printf("| %7d | %5llu us |\n", i, std::chrono::duration_cast<std::chrono::microseconds>(end-start).count());
      }
    } catch (af::exception& e) {

        fprintf(stderr, "%s\n", e.what());
        throw;
    }

    return 0;
}
```

### CODE REVIEW

File: src/api/unified/symbol_manager.hpp [hunk 1]
```diff
@@ -54,13 +57,18 @@ class AFSymbolManager {
             if (!activeHandle) {
                 UNIFIED_ERROR_LOAD_LIB();
             }
+
             typedef af_err(*af_func)(CalleeArgs...);
-            af_func funcHandle;
+            static std::unordered_map<const char*, std::array<af_func, NUM_BACKENDS>> funcHandles;
+            af_func& funcHandle = funcHandles[symbolName][getActiveBackend()];
```
Comment (pavanky) [inline src/api/unified/symbol_manager.hpp]: Does it always return `NULL` when it is not found ?

File: src/api/unified/symbol_manager.hpp [hunk 2]
```diff
@@ -54,13 +57,18 @@ class AFSymbolManager {
             if (!activeHandle) {
                 UNIFIED_ERROR_LOAD_LIB();
             }
+
             typedef af_err(*af_func)(CalleeArgs...);
-            af_func funcHandle;
+            static std::unordered_map<const char*, std::array<af_func, NUM_BACKENDS>> funcHandles;
```
Comment (pavanky) [inline src/api/unified/symbol_manager.hpp]: Move the array outside ?

File: src/api/unified/symbol_manager.hpp (hunk 1 above)
Comment (umar456) [inline src/api/unified/symbol_manager.hpp]: I am 96.3% sure it does.

File: src/api/unified/symbol_manager.hpp (hunk 2 above)
Comment (umar456) [inline src/api/unified/symbol_manager.hpp]: 👍

File: src/api/unified/symbol_manager.hpp [hunk 3]
```diff
@@ -59,14 +59,14 @@ class AFSymbolManager {
             }
 
             typedef af_err(*af_func)(CalleeArgs...);
-            static std::unordered_map<const char*, std::array<af_func, NUM_BACKENDS>> funcHandles;
-            af_func& funcHandle = funcHandles[symbolName][getActiveBackend()];
+            static std::array<std::unordered_map<const char*, af_func>, NUM_BACKENDS> funcHandles;
+            af_func& funcHandle = funcHandles[getActiveBackend()][symbolName];
```
Comment (pavanky) [inline src/api/unified/symbol_manager.hpp]: Is `getActiveBackend` returning the enum or is it returning the index ?

Comment (umar456) [inline src/api/unified/symbol_manager.hpp]: oops. It's a flag. Fixing.

File: src/api/unified/symbol_manager.hpp [hunk 4]
```diff
@@ -37,6 +37,14 @@ const int NUM_ENV_VARS = 2;
                     "for instructions to set up environment for Unified backend.", \
                     AF_ERR_LOAD_LIB)
 
+static int backend_index(af::Backend be) {
+    switch (be) {
+        case AF_BACKEND_CPU: return 0;
+        case AF_BACKEND_CUDA: return 1;
+        case AF_BACKEND_OPENCL: return 2;
+        default: return -1;
```
Comment (pavanky) [inline src/api/unified/symbol_manager.hpp]: Hmm, we have something called `AF_BACKEND_DEFAULT`. Please check if it works with that.

**Comment by pavanky** [comment]
@umar456 Alternative suggestion: Instead of querying from an `unordered_map` each time, can't the function pointer be cached inside the caller function ? 

This reduces the lookup time from the map. You will need change the definition of `CALL` macros.

File: src/api/unified/symbol_manager.hpp (hunk 4 above)
Comment (umar456) [inline src/api/unified/symbol_manager.hpp]: The [SymbolManager](https://github.com/arrayfire/arrayfire/blob/devel/src/api/unified/symbol_manager.cpp#L158) and [setBackend](https://github.com/arrayfire/arrayfire/blob/devel/src/api/unified/symbol_manager.cpp#L203) function sets the activeBackend to one of those backends.

**Comment by umar456** [comment]
I tried that first but the call function's template parameters are based on the type of arguments and not the function name itself. I could change the macro so that we only have an array instead of a map but I didn't want to break anything.

I can change it if you think its worthwhile.

**Comment by pavanky** [comment]
We can still pass the arguments, but instead of calling the function, it will just return the function pointer. The call can be made externally. No reason to cache it inside the `call` function.

**Comment by umar456** [comment]
build arrayfire windows ci

**Pull request merged**